\name{plots}
\alias{pdistr}
\alias{volcano}
\alias{exprs.heatmap}
\title{
    Visualization of gene expression
}
\description{
    Visualization of differential gene expression via 
    heatmap, p-value histogram and volcano plot (fold change vs. p-value).
}
\usage{
    pdistr( p )
    volcano( fc, p )
    exprs.heatmap( expr, grp, scale.rows=TRUE, log.thresh=100 )
}
\arguments{
    \item{p}{
        Numeric vector of p-values for each gene.
    }
    \item{fc}{
        Numeric vector of fold changes (typically on log2 scale).
    }
    \item{expr}{
        Expression matrix. Rows correspond to genes, columns to samples.    
    }
    \item{grp}{
        *BINARY* group assignment for the samples. 
        Use '0' and '1' for unaffected (controls) and affected (cases) 
        samples, respectively.
    }
    \item{scale.rows}{
        Should rows of the expression matrix be scaled for better visibility
        of expression differences between sample groups? Defaults to TRUE.
    }
    \item{log.thresh}{
        Threshold for log2-transformation of the expression matrix.
        Particularly useful for heatmap visualization of RNA-seq read count data,
        where the max and the min of the expression matrix typically differ by 
        several orders of magnitude.
        If the difference between min and max of the expression matrix is greater
        than the indicated threshold, log2-transformation is applied. 
    }
}
\value{
    None, plots to a graphics device.
}
\author{
    Ludwig Geistlinger <Ludwig.Geistlinger@bio.ifi.lmu.de>
}
\seealso{
    \code{\link{de.ana}} for differential expression analysis,
    \code{\link{heatmap}} and \code{\link{truehist}} for generic plotting. 
}
\examples{
    # (1) simulating expression data: 100 genes, 12 samples
    eset <- make.example.data(what="eset") 
    
    # plot heatmap
    exprs.heatmap(expr=exprs(eset), grp=as.factor(pData(eset)$GROUP))

    # (2) DE analysis
    eset <- de.ana(eset)
    pdistr(fData(eset)$ADJ.PVAL)
    volcano(fc=fData(eset)$FC, p=fData(eset)$ADJ.PVAL)
}
